File: Python/Math-Array-String/23.py
# 507. Perfect Number

# Only 5 perfect Number exist until 10^8
# class Solution:
#     def checkPerfectNumber(self, num: int) -> bool:
#         return num in [6,28,496,8128,33550336]

class Solution:
    def checkPerfectNumber(self, num: int) -> bool:
        if num <= 1:
            return False
        
        divisor_sum = 1  # Start with 1 because 1 is a divisor of every number
        # Check divisors up to sqrt(n)
        for i in range(2, int(num**0.5) + 1):
            if num % i == 0:  # If i is a divisor
                divisor_sum += i
                if i != num // i:  # Add the complement divisor if it's not the same
                    divisor_sum += num // i
        
        return divisor_sum == num


sol = Solution()
output = sol.checkPerfectNumber(28)
print(output)

File: JavaScript/Math-Array-String/23.js
// 507. Perfect Number

/**
 * @param {number} num
 * @return {boolean}
 */
var checkPerfectNumber = function(num) {
    if (num <= 1) return false;

    let divisorSum = 1; 

    for (let i = 2; i <= Math.sqrt(num); i++) {
        if (num % i === 0) { 
            divisorSum += i;
            if (i !== num / i) { 
                divisorSum += num / i;
            }
        }
    }

    return divisorSum === num;
};


// Example usage
console.log(checkPerfectNumber(28)); // true (Perfect number: 1 + 2 + 4 + 7 + 14 = 28)
console.log(checkPerfectNumber(6));  // true (Perfect number: 1 + 2 + 3 = 6)
console.log(checkPerfectNumber(12)); // false (Sum of divisors: 1 + 2 + 3 + 4 + 6 = 16)